helix-view/src/register.rs

```diff
@@ -0,0 +1,235 @@
+        match name {
+            '_' => Some(Box::new(iter::empty())),
+            '#' => {
+                let (view, doc) = current_ref!(editor);
+                let selections = doc.selection(view.id).len();
+                // ExactSizeIterator is implemented for Range<usize> but
+                // not RangeInclusive<usize>.
+                Some(Box::new(
+                    (0..selections).map(|i| i.saturating_add(1).to_string().into()),
+                ))
+            }
+            '.' => {
+                let (view, doc) = current_ref!(editor);
+                let text = doc.text().slice(..);
+                Some(Box::new(doc.selection(view.id).fragments(text)))
+            }
+            '%' => {
+                let doc = doc!(editor);
+
+                let path = doc
+                    .path()
+                    .as_ref()
+                    .map(|p| p.to_string_lossy())
+                    .unwrap_or_else(|| SCRATCH_BUFFER_NAME.into());
+
+                Some(Box::new(iter::once(path)))
+            }
+            '*' | '+' => Some(read_from_clipboard(
+                self.clipboard_provider.as_ref(),
+                self.inner.get(&name),
+                match name {
+                    '*' => ClipboardType::Clipboard,
+                    '+' => ClipboardType::Selection,
+                    _ => unreachable!(),
+                },
+            )),
+            _ => self
+                .inner
+                .get(&name)
+                .map(|values| Box::new(values.iter().map(Cow::from).rev()) as RegisterValues),
+        }
+    }
+
+    pub fn write(&mut self, name: char, mut values: Vec<String>) -> Result<()> {
+        match name {
+            '_' => Ok(()),
+            '#' | '.' | '%' => Err(anyhow::anyhow!("Register {name} does not support writing")),
+            '*' | '+' => {
+                self.inner.insert(name, values);
+                self.clipboard_provider.set_contents(
+                    self.inner[&name].join(NATIVE_LINE_ENDING.as_str()),
+                    match name {
+                        '*' => ClipboardType::Clipboard,
+                        '+' => ClipboardType::Selection,
+                        _ => unreachable!(),
+                    },
+                )
+            }
+            _ => {
+                values.reverse();
+                self.inner.insert(name, values);
+                Ok(())
+            }
+        }
+    }
+
+    pub fn push(&mut self, name: char, value: String) -> Result<()> {
+        match name {
+            '_' => Ok(()),
+            '#' | '.' | '%' => Err(anyhow::anyhow!("Register {name} does not support pushing")),
+            '*' | '+' => {
+                self.inner.entry(name).or_insert_with(Vec::new).push(value);
+                self.clipboard_provider.set_contents(
+                    self.inner[&name].join(NATIVE_LINE_ENDING.as_str()),
```

Comment:
This doesn't work correctly if the clipboard has changed outside of helix altough that might be ok. I am not sure how we want to handle that case?

Reply:
I opted to use the `read_from_clipboard` helper and then push/prepend to the values returned from that